[export] Add run_decomposition() function to ExportedProgram

angelayi opened the pull request:
Summary:
https://docs.google.com/document/d/1QJJEGnj2nHGPODlw38BEG3KLLCOTfdOVjPrNQbz_LM8/edit#bookmark=id.lp80wfshq130

`exported_program.run_decompositions(decomposition_table)` will optionally take a decomposition table, and run decompositions on the exported program, returning a new exported program. By default we will run the Core ATen decomposition table.

Splitting up this diff with the following one (D49742989) to make migrating Executorch easier:
1. Land this diff
1. Wait for a pytorch nightly to include this diff
1. Update executorch's pytorch nightly
1. Land the following diff to have export() return no decomps

Test Plan: Tested in following diff

Differential Revision: D49743208

gmagogsfm commented on test/export/test_export.py:
```diff
@@ -1449,5 +1449,61 @@ def forward(self, x):
+        ).run(core_aten_ep.graph_module.code)
+        self.assertTrue(torch.allclose(core_aten_ep(*inp), m(*inp)))
+
+    def test_export_decomps_dynamic(self):
```
I am all for having more tests. Just a bit curious why dynamic shape affects run-decomposition.

angelayi replied:
I just wanted to check that the dynamic shapes were preserved through aot export.

angelayi commented on torch/export/exported_program.py:
```diff
@@ -422,43 +423,151 @@ def module(self) -> torch.nn.Module:
+                    new_range_constraints, new_equality_constraints
+                )
+            )
+        exported_program = lift_constant_tensor_pass(exported_program)
```
I'm not sure if this is 100% needed. Basically, if it's possible that some decomposed operator introduces a constant tensor, then we will need to lift it.

angelayi commented on torch/export/exported_program.py:
```diff
@@ -422,43 +423,151 @@ def module(self) -> torch.nn.Module:
+
+        if len(new_range_constraints) > 0 or len(new_equality_constraints) > 0:
+            exported_program = exported_program._transform(
+                _AddRuntimeAssertionsForInlineConstraintsPass(
```
I'm also not sure if this is 100% needed. If it's possible that some decomposed operator introduces an unbacked symint, then we will also need to add assertions for it 😅

angelayi committed:
[export] Add run_decomposition() function to ExportedProgram (#110236)

Summary:

https://docs.google.com/document/d/1QJJEGnj2nHGPODlw38BEG3KLLCOTfdOVjPrNQbz_LM8/edit#bookmark=id.lp80wfshq130

`exported_program.run_decompositions(decomposition_table)` will optionally take a decomposition table, and run decompositions on the exported program, returning a new exported program. By default we will run the Core ATen decomposition table.

Splitting up this diff with the following one (D49742989) to make migrating Executorch easier:
1. Land this diff
1. Wait for a pytorch nightly to include this diff
1. Update executorch's pytorch nightly pin
1. Land the following diff to have export() return no decomps + updates executorch code to use the run_decomps()

Test Plan: Tested in following diff

Reviewed By: tugsbayasgalan

Differential Revision: D49743208
